src/js_native.rs

```diff
@@ -0,0 +1,147 @@
+//! Helpers for exposing native functions to JavaScript.
+
+/// Wrap a `JSNative` function with a catch unwind to ensure we don't panic and
+/// unwind across language boundaries, which is undefined behavior.
+#[macro_export]
+macro_rules! js_native_no_panic {
+    (
+        $( #[$attr:meta] )*
+        pub fn $name:ident ( $( $arg_name:ident : $arg_ty:ty $(,)* )* ) -> bool {
+            $( $body:tt )*
+        }
+    ) => {
+        $( #[$attr] )*
+        pub extern "C" fn $name( $( $arg_name : $arg_ty , )* ) -> bool {
+            match ::std::panic::catch_unwind(move || {
+                $( $body )*
+            }) {
+                Ok(b) => b,
+                Err(_) => {
+                    eprintln!(
+                        concat!(
+                            "error: JSNative '",
+                            stringify!($name),
+                            "' panicked: must not panic across FFI boundaries!"
+                        )
+                    );
+                    false
+                }
+            }
+        }
+    }
+}
+
+macro_rules! len_idents {
+    () => {
+        0
+    };
+    ( $x:ident $( $xs:ident )* ) => {
+        1 + len_idents!( $( $xs )* )
+    }
+}
+
+/// Define a native Rust function that that can be exposed to JS as a
+/// `JSNative`.
+///
+/// All parameter types must implement the
+/// `js::conversions::FromJSValConvertible<Config = ()>` trait.
+///
+/// The return type must implement `js::conversions::ToJSValConvertible`.
```

Comment:
I suppose we'll have to provide impls for a range of builtin types eventually. Obviously it makes sense to do that once we have a bit more experience with how this feels.

Reply:
Indeed.